File: tests/basic/reference_inline.ml
open Optitrust
open Target

let _ = Run.doc_script_cpp (fun _ ->
  !! Variable_basic.inline [cVarDef "x"];
  )
"
int main() {
  const int& x = 3;
  int y = x;
  y++;
}
"

let _ = Run.script_cpp ( fun _ ->
  (* inline at one specific occurence *)
  !! Variable_basic.unfold ~at:[cVarDef "r1"] [cVarDef "y"];
  !! Variable_basic.unfold ~at:[cVarDef "r3"] [cVarDef "b"];
  (* inline at all occurences and delete the reference definition *)
  !! Variable_basic.inline [cVarDef "a"];
  (* inline a reference to a matrix row *)
)

File: tests/basic/reference_inline.cpp

int main() {
   // inline reference on an integer
   int x = 2;
   const int& y = x;
   int r1 = y + y;

   // inline reference on array cell
   int t[2] = { 4, 5 };
   const int& a = t[1];
   int r2 = a + a;

   // inline reference on matrix cell
   int m[3][3];
   const int& b = m[1][1];
   int r3 = b + b;

   // inline reference on matrix row
   const auto& v = m[0]; // LATER: what is the type of v?
   int r4 = v[2];
   
}

